src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/BatchUpdateTest.java

```diff
@@ -287,6 +287,39 @@ void beginTransaction(Invocation inv) throws DatabaseEngineRuntimeException {
         assertEquals("check that only 1 transaction was really executed", 1, transactions.get());
     }
 
+    /**
+     * Tests that there is no race condition between the {@link AbstractBatch#destroy()} and {@link AbstractBatch#run()}
+     * methods.
+     * This is a regression test for PULSEDEV-18139, where a race condition was causing the scheduler to attempt to
+     * call run while it another thread was already inside `destroy` but had not yet called shutdown on the scheduler.
+     * Since those two methods were synchronized, the `run` would not finish while destroy was waiting for all tasks in
+     * the Executor to finish.
+     *
+     *
+     * @since 2.1.10
+     * @throws DatabaseEngineException If the operations on the engine fail.
+     */
+    @Test(timeout = 30000)
+    public void testBatchRunDestroyRace() throws DatabaseEngineException {
+        final DbEntity entity = dbEntity()
+                .name("TEST")
+                .addColumn("COL1", INT)
+                .addColumn("COL2", BOOLEAN)
+                .addColumn("COL3", DOUBLE)
+                .addColumn("COL4", LONG)
+                .addColumn("COL5", STRING).build();
+
+        engine.addEntity(entity);
+
+
+        for (int i = 0; i < 40; i++) {
+            final MockedBatch batch = MockedBatch.create(engine, "test", 5, 10L, 50000);
```

Comment:
Add a comment expliciting that the "wait" timeout defined here is larger than the test timeout. Thus, if the regression was ever hit, it would generate a test timeout.